cmd/podman/images/scp.go

```diff
@@ -0,0 +1,330 @@
+			return errors.Wrapf(define.ErrInvalidArg, "Renaming of an image is currently not supported")
+		}
+		scpOpts.Save.Format = "oci-archive"
+		abiErr := abiEng.Save(context.Background(), args[0], tags, scpOpts.Save) // save the image locally before loading it on remote, local, or ssh
+		if abiErr != nil {
+			errors.Wrapf(abiErr, "could not save image as specified")
+		}
+		rep, err := abiEng.Load(context.Background(), scpOpts.Load)
+		if err != nil {
+			return err
+		}
+		fmt.Println("Loaded images(s): " + strings.Join(rep.Names, ","))
+		err = os.Remove(scpOpts.Load.Input)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// execCommand takes ssh information and an image file path on the client, connects via ssh, executes
+// an scp command and then calls podman image load on the remote machine given the new scp filepath
+func execCommand(iden string, uri *urlP.URL, imageFileIn string, imageFileOut string, image string, tag string, LoadSave bool) (string, error) {
+	cfg, err := connection.Validate(uri, iden)
+	if err != nil {
+		return " ", err
+	}
+	dial, err := ssh.Dial("tcp", uri.Host, cfg) // dial the client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to connect")
+	}
+	defer dial.Close()
+
+	session, err := dial.NewSession() // connect to the dialed client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create new ssh session on %q", uri.Host)
+	}
+	defer session.Close()
+	if err != nil {
+		return "", err
+	}
+
+	if LoadSave { // if we are loading to the remote/ssh client
+		f, _ := os.Open(imageFileIn) // open image file to copy
+		if err != nil {
+			return "", err
+		}
+		defer f.Close()
+		n, err := scpD.CopyTo(dial, imageFileIn, imageFileOut)
+		if err != nil {
+			fmt.Println(n, " Bytes copied before error")
+			return " ", err
+		}
+
+		podman := os.Args[0]
+		if v, found := os.LookupEnv("PODMAN_BINARY"); found {
+			podman = v // get path to podman bin
+		}
+		run := ""
+		if tag != "" {
+			return "", errors.Wrapf(define.ErrInvalidArg, "Renaming of an image is currently not supported")
+		}
+		run = podman + " image load --input=" + imageFileOut + ";rm " + imageFileOut // run ssh image load of the file copied via scp
+		var buffer bytes.Buffer
+		session.Stdout = &buffer
+		if err := session.Run(run); err != nil { // run the command declared above
+			return "", err
+		}
+		return buffer.String(), nil // return the output of image load on the machine, should be "Loaded Image(s)..."
+	} // else we are loading from the ssh client
+	podman := os.Args[0]
+	if v, found := os.LookupEnv("PODMAN_BINARY"); found {
+		podman = v // get path to podman bin
+	}
+	run := ""
+	if tag != "" {
+		return "", errors.Wrapf(define.ErrInvalidArg, "Renaming of an image is currently not supported")
+	}
+	run = podman + " image save " + imageFileOut + " --output=" + imageFileIn // run ssh image load of the file copied via scp. Files are reverse in thie case...
+	var buffer bytes.Buffer
+	session.Stdout = &buffer
```

Comment:
I think for now capturing stdErr and returning it with the session error makes sense. Keeping this in the image realm vs the system/connection realm makes more sense to me

Reply:
Ideally, the shared code, if it can be formed into a coherent whole, should be somewhere in `pkg/ssh` rather than `system/connection/add`.

But at this point this is already calling directly into `system/connection/add`, so one more dependency like that is not much of a problem — whereas the code duplication is a clear benefit: adding the `Stderr` improvement here is nice, but sharing the code to benefit the pre-existing implementation is better.

Reply:
this command is not run in `system/connection/add`. I feel like putting this command in that file dilutes it since that file is only supposed to be for creating a new connection not executing a podman image command.

Reply:
Yes, s/c/add does not run _this_ command but it runs _a_ command. So the `dial.NewSession`+the two `bytes.Buffer` values + inclusion of stderr on failure could all be encapsulated into a nice little `execRemoteCommand` function. Especially now that `execCommand` needs to run up to three (`mktemp`, `save`, `rm`).

---

If you feel it doesn’t belong into s/c/add, move it into `system/connection/shared.go`.